Optimized autoraster

vshekar opened the pull request:
Initial commit for optimized autoraster, work in progress 

Do the ophyd devices defined need to be in mxtools or stay here? 
What is the criteria of having devices in mxtools? As there are multiple devices in start_bs.py and setenergy_lsdc.py

In this PR there are 2 major contributions:
1. A bluesky plan to align the pin on mount using the top view camera as part of state transitions from SE -> TA -> SA
2. A bluesky plan to replace autoRasterLoop that uses an ML model to identify a bounding box around the sample and run 2 2D rasters around the face on and orthogonal positions

JunAishima commented on bluesky_env/plans/top_view.py:
```diff
@@ -0,0 +1,133 @@
+        omega_min = np.linspace(omegas[0], omegas[-1], sample)[
+            b_splines.argmin()
+        ]
+        print(f"SPLINES / {omega_min}")
```
log as debug?

vshekar replied:
Added to logger.debug instead of print in 63ccbc857f8414e0fe27da5ee59e7cb547740872

JunAishima commented on bluesky_env/devices/top_align.py:
```diff
@@ -0,0 +1,343 @@
+                ("tiff.nd_array_port", "CV1"),
+            ]
+        )
+        if self.cam_mode.get() == "coarse_align":
```
some docs or other info about the `cam_mode`s, as well as having constants for the strings would help clean up the code and reduce issues with setting the values.

vshekar replied:
Added enum for cam modes and doc strings in 1608ff2f2a52dea4959c622a2f9cd604414f0f35

JunAishima commented on daq_macros.py:
```diff
@@ -211,9 +213,54 @@ def changeImageCenterHighMag(x,y,czoom):
 
 def autoRasterLoop(currentRequest):
+    global sample_detection, autoRasterFlag, max_col
+    RE(loop_center_plan())
```
we should start thinking about how to better handle running plans within these functions. just thinking ahead to the future where the highest level function gets executed in the Queueserver (or its successors), and this won't be able to run because you can't have a RE(call) within a plan.

JunAishima replied:
I see this has been changed. resolving as done.

JunAishima commented on daq_macros.py:
```diff
@@ -2201,14 +2248,19 @@ def runRasterScan(currentRequest,rasterType=""): #this actually defines and runs
 
 def gotoMaxRaster(rasterResult,multiColThreshold=-1,**kwargs):
-  global autoVectorCoarseCoords,autoVectorFlag
+  global autoVectorCoarseCoords,autoVectorFlag, max_col
```
gotoMaxRaster is a real mess. this needs to be refactored (as we had seen with the multiCol upgrade attempts before)

vshekar replied:
I will attempt to refactor it in the next PR

JunAishima commented on embl_robot.py:
```diff
@@ -207,7 +207,7 @@ def preMount(self, gov_robot, puckPos, pinPos, sampID, **kwargs):
             kwargs['prefix1'] = prefix1
             kwargs['prefix90'] = prefix90
-            top_view.topViewSnap(prefix1,getBlConfig("visitDirectory")+"/pinAlign",1,acquire=0)
+            # top_view.topViewSnap(prefix1,getBlConfig("visitDirectory")+"/pinAlign",1,acquire=0)
```
any reason this can't just be removed?

vshekar replied:
Currently keeping this just for FMX (code now has a beamline==FMX check) until their top cameras are in place until then the old top_view can do the initial alignment

JunAishima commented on daq_macros.py:
```diff
@@ -229,36 +230,44 @@ def run_top_view_optimized():
+      if gov_status.success:
+        yield from detect_loop(sample_detection)
+      else:
+        print("could not trasition to SA")
```
transition (sp)

vshekar commented:
> I can't review this very thoroughly, but there are various comments and changes that need to be addressed here.

Understandable, I'm in the process of addressing issues with the bluesky plan itself (error handling, correctness of motions etc.). Then I will address your initial comments before I ask for a more detailed review

vshekar commented on bluesky_env/devices/zebra.py:
```diff
@@ -0,0 +1,698 @@
+print(f"Loading {__file__}")
```
The beginning few devices are the same, a few devices have a custom init function and 4 position capture devices are new. Although I don't think we need to include it in this PR, I'll keep an eye on it and make sure if we are adding more ophyd objects, I'll refer to that library and include it. 

In that future PR I'll make it a point to integrate the common zebra devices from this package and get rid of the ones in LSDC

vshekar commented on bluesky_env/devices/auto_center.py:
```diff
@@ -0,0 +1,211 @@
+    jpeg = Cpt(
+        JPEGPluginWithFileStore,
+        "JPEG1:",
+        write_path_template="/nsls2/data/amx/legacy/topcam",
```
Streaming images directly to the endpoint is a work in progress, probably in a future update. We wanted to save the images in one place accessible only to us for later training, ideally the users don't need this. Any suggestions for a neutral location for these files besides the one above?

JunAishima replied:
if the users don't need to see this, it's good enough

vshekar commented on bluesky_env/devices/auto_center.py:
```diff
@@ -0,0 +1,211 @@
+        self.cv1.outputs.read_attrs = ["output1", "output2", "output3"]
+        self.cam_mode.subscribe(self._update_stage_sigs, event_type="value")
+
+    def _update_stage_sigs(self, *args, **kwargs):
```
Potentially, although tests at FMX without changes to these values were successful

JunAishima replied:
good for now.

vshekar commented on bluesky_env/devices/auto_center.py:
```diff
@@ -0,0 +1,211 @@
+
+class LoopDetector(Device):
+    url = Cpt(
+        Signal, value=f'{os.environ["SAMPLE_DETECT_URL"]}/predict', kind='config'
```
Model is still research code but will release on github and torch.hub once everything is stabilized. Will add comments on how to use URL. Future updates will include repo details

vshekar replied:
Added comments related to URL in 0b436cffb19d1a22ffdbaee4312e9f8f4a1247fc

vshekar commented on bluesky_env/devices/auto_center.py:
```diff
@@ -0,0 +1,211 @@
+        super().stage(*args, **kwargs)
+
+
+class WorkPositions(Device):
```
Moved to generic devices file in 461484cce3b8a9d1dc22a669560e25c0bcbdaac9

vshekar commented on bluesky_env/devices/auto_center.py:
```diff
@@ -0,0 +1,211 @@
+
```
Changed bluesky_env to mxbluesky in 40baf5b98e8403948812ecfbf3a69ae573d7777f

vshekar commented on bluesky_env/devices/zebra.py:
```diff
@@ -0,0 +1,698 @@
+            }
+        }
+
+zebra1 = Zebra('XF:17IDB-ES:AMX{Zeb:1}:', name='zebra1')
```
These devices aren't used anywhere, removed in c6082466618b86020f9ac8595eb0ea3628cef2b7

vshekar commented on bluesky_env/devices/auto_center.py:
```diff
@@ -0,0 +1,211 @@
+                    ("roi4.size.y", self.roi1.size.y.get()),
+                ]
+            )
+        elif self.cam_mode.get() == "two_click":
```
The device and plan mimics two click centering where the center of the sample is moved to the center of the beam and then its orthogonal position is also moved to the center

vshekar commented on bluesky_env/devices/top_align.py:
```diff
@@ -0,0 +1,343 @@
+        write_path_template="/nsls2/data/amx/legacy/topcam",
+    )
+    cam_mode = Cpt(Signal, value=None, kind="config")
+    pix_per_um = Cpt(Signal, value=0.164, doc="pixels per um")
```
This is true

vshekar commented on bluesky_env/plans/loop_detection.py:
```diff
@@ -0,0 +1,121 @@
+    yield from bp.count([two_click_low], 1)
+    loop_detector.filename.set(two_click_low.jpeg.full_file_name.get())
+
+    """
```
Not a docstring, meant to comment out some code. Removed in 9f39bf7b5ebbca29204e5f96ee226ce0a2ea3238

vshekar commented on bluesky_env/plans/loop_detection.py:
```diff
@@ -0,0 +1,121 @@
+    omega: float = gonio.o.user_readback.get()
+    sample_detection["face_on_omega"] = omega
+
+    d = np.pi/180
```
Fixed in b67e7594f6511b88cfef1f3a88358598c57f06a5

vshekar commented on bluesky_env/plans/loop_detection.py:
```diff
@@ -0,0 +1,121 @@
+    yield from bps.mvr(gonio.pz, -real_z)
+
+    # The sample has moved to the center of the beam (hopefully), need to update co-ordinates
+    box_coords_face[0]
```
Stray statement, removed in b67e7594f6511b88cfef1f3a88358598c57f06a5

vshekar commented on bluesky_env/devices/top_align.py:
```diff
@@ -0,0 +1,343 @@
+        super().stage(*args, **kwargs)
+
+
+class ZebraMXOr(Zebra):
```
Moved device to zebra device file in 8d2de437fd3c769275546e4fc24126d276dfedcd

vshekar commented on bluesky_env/devices/top_align.py:
```diff
@@ -0,0 +1,343 @@
+    stats4 = Cpt(StatsPlugin, "Stats4:")
+    stats5 = Cpt(StatsPlugin, "Stats5:")
+
+class EpicsMotorSPMG(EpicsMotor):
```
@dalekreitler-bnl do you know what this epics motor does?

vshekar commented on bluesky_env/plans/loop_detection.py:
```diff
@@ -0,0 +1,121 @@
+
+    if len(box_coords_face) != 4:
+        logger.exception("Exception during loop detection plan. Face on loop not found")
+        sample_detection["large_box_width"] = 630 * 2 * two_click_low.pix_per_um.get()
```
Using configurable values in 831f5d8bd3e7b58fa0c88ea2c46cd35f3a62427f

vshekar commented on daq_macros.py:
```diff
@@ -232,6 +232,42 @@ def changeImageCenterHighMag(x,y,czoom):
 def run_top_view_optimized():
     RE(topview_optimized())
 
+def run_on_mount_option(sample_id):
```
This function is used to provide interactive users the ability to 
1. Do nothing on mount 
2. Only center sample
3. Center sample and take 2 orthogonal rasters of the sample

vshekar commented:
Thorough testing of code done at both beamlines. PR ready for full review

JunAishima reviewed (commented):
My main concern is whether this pull request has code that is maintainable and extensible in the long-term. This may indicate that we should do an architecture review to understand how to enable new methods such as this and the MD2 collections on NYX.

vshekar commented on daq_lib.py:
```diff
@@ -233,11 +233,11 @@ def getRobotConfig():
 
 
-def setRobotGovState(stateString):
+def setRobotGovState(stateString, wait=True):
```
Initially edited the existing code, replaced and removed setRobotGovState() with gov_lib.setGovRobot() for all references in the code: 94f9eeb0a8e14b83f8ab02e8093a920e5ea24cee

vshekar commented on daq_macros.py:
```diff
@@ -211,9 +216,115 @@ def changeImageCenterHighMag(x,y,czoom):
+def loop_center_plan():
+    global sample_detection
+    if gov_robot.state.get() == 'M':
+      bps.sleep(15)
```
I will keep it in for now with a comment to indicate the reason for the sleep. If this same pattern starts creeping into other plans I will create a decorator

vshekar commented on daq_macros.py:
```diff
@@ -2260,6 +2377,58 @@ def gotoMaxRaster(rasterResult,multiColThreshold=-1,**kwargs):
+        ceiling = 0
+        # Loop through all the cells that belong to the column corresponding to max_col
+        for index in indices:
+          try:
```
A similar loop happens in this function with does the exact same thing but it handles multicol cases in there as well. I will refactor this function because it is a pain for me to add or modify it due to the number of loops and if conditions

vshekar replied:
Fixed by refactoring function here: fdf7306cd9bde3739dc4dcdd103f6ed1fcdc1f8b

vshekar commented on mxbluesky/devices/zebra.py:
```diff
@@ -0,0 +1,700 @@
+print(f"Loading {__file__}")
```
Only the initial devices are similar, the actual Zebra device and its parent classes are fairly different. I think its also best to keep the common devices in case of breaking changes upstream

vshekar commented on mxbluesky/devices/top_align.py:
```diff
@@ -0,0 +1,401 @@
+
+class TopAlignerFast(TopAlignerBase):
+
+    zebra = Cpt(ZebraMXOr, "XF:17IDB-ES:AMX{Zeb:2}:")
```
Changed here: https://github.com/NSLS-II/lsdc/pull/355/commits/973088df62fabb72ba3969b1eb44207348c2a636

vshekar commented on mxbluesky/devices/auto_center.py:
```diff
@@ -0,0 +1,209 @@
+                ("cam.acquire_period", 1),
+            ]
+        )
+        if self.cam_mode.get() == "centroid":
```
Unlikely to change and tested at both beamlines, but even if it did where do you propose they be stored

vshekar commented on mxbluesky/devices/auto_center.py:
```diff
@@ -0,0 +1,209 @@
+    jpeg = Cpt(
+        JPEGPluginWithFileStore,
+        "JPEG1:",
+        write_path_template="/nsls2/data/amx/legacy/topcam",
```
Most likely be compiled into an h5 file for further training of the model

vshekar commented:
> My main concern is whether this pull request has code that is maintainable and extensible in the long-term. This may indicate that we should do an architecture review to understand how to enable new methods such as this and the MD2 collections on NYX.

As far as I understand it this code is pretty modular, devices and plans related to this are in their own python modules. Options are available via the GUI to select whether to use this feature or not. start_bs.py specifically initializes devices for each beamline. The 2 plans developed as part of this PR has only one way to use it `daq_macros.run_top_view_optimized()` and `daq_macros.run_loop_center_plan()` both of which can be gated behind an `if beamline == 'amx' or beamline == `fmx` statement. Or add conditions within that one function if nyx wants to do things their own way.

Are there specific long term concerns you have with this code? I can attempt to address them in a future PR

JunAishima commented:
I think I went through a picky phase when doing the review and should have kept in mind: "premature optimization". I think you will see over time where the parts that need more flexibility due to changing parameters, and changing hardware/etc, so you should refactor then.
